base/iterators.jl

```diff
@@ -69,6 +70,77 @@ eltype{I}(::Type{Enumerate{I}}) = Tuple{Int, eltype(I)}
 iteratorsize{I}(::Type{Enumerate{I}}) = iteratorsize(I)
 iteratoreltype{I}(::Type{Enumerate{I}}) = iteratoreltype(I)
 
+struct IndexValue{I,A<:AbstractArray}
+    data::A
+    itr::I
+end
+
+"""
+    enumerate(IndexLinear(), A)
+    enumerate(IndexCartesian(), A)
+    enumerate(IndexMethod(A), A)
+
+An iterator that accesses each element of the array `A`, returning
+`(i, x)`, where `i` is the index for the element and `x = A[i]`.  This
+is similar to `enumerate(A)`, except `i` will always be a valid index
+for `A`.  Like `enumerate`, the bounds-check that would otherwise
+occur for `A[i]` is omitted.
+
+Specifying specifying `IndexLinear()` ensures that `i` will be an
+integer; specifying `IndexCartesian()` ensures that `i` will be a
+`CartesianIndex`; specifying `IndexMethod(A)` chooses whichever has
+been defined as the native indexing method for array `A`.
+
+```jldoctest
+julia> A = ["a" "d"; "b" "e"; "c" "f"];
+
+julia> for (index, value) in enumerate(IndexMethod(A), A)
+           println("\$index \$value")
+       end
+1 a
+2 b
+3 c
+4 d
+5 e
+6 f
+
+julia> S = view(A, 1:2, :);
+
+julia> for (index, value) in enumerate(IndexMethod(S), S)
+           println("\$index \$value")
+       end
+CartesianIndex{2}((1, 1)) a
+CartesianIndex{2}((2, 1)) b
+CartesianIndex{2}((1, 2)) d
+CartesianIndex{2}((2, 2)) e
+```
+
+Note that for a vector `v`, `enumerate(v)` returns `i` as a *counter*
+(always starting at 1), whereas `enumerate(IndexLinear(), v)` returns
+`i` as an *index* (starting at the first index of `v`, which may or
+may not be 1).
```

Comment:
I don't think there's a reason to special-case vectors in the base documentation at all. Even if you decide not to change the offset-ness of a linear index in an `OffsetMatrix`, another package could do so.